keps/sig-node/20181106-in-place-update-of-pod-resources.md

```diff
@@ -153,6 +153,18 @@ Additionally, Kubelet is authorized to update PodSpec, and NodeRestriction
 admission plugin is extended to limit Kubelet's update access only to Pod's
 ResourcesAllocated field for CPU and memory resources.
 
+<<[UNRESOLVED]>>
+Why is "allocated" spec?  The need for special admission control is a strong
+smell.  Why doesn't this belong in status?  If there's a compelling reason,
+this KEP should say so.
+<<[/UNRESOLVED]>>
```

Comment:
Summarizing [our conversation](https://github.com/kubernetes/kubernetes/pull/92127#discussion_r448523145) from https://github.com/kubernetes/kubernetes/pull/92127"
- API conventions require [Status to be reconstructible by observation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status) if lost
- If ResourcesAllocated was in Status, and it were lost or erased, as of today it cannot be reconstructed purely based on observations on the node - memory request in particular.
- We discussed it at length in sig-node, explored the alternative of node-local checkpointing instead of Spec change, and eventually felt Spec.ResourcesAllocated was the best option - KEP discussion: https://github.com/kubernetes/enhancements/pull/686#discussion_r301865371

Reply:
I tried to read that comment thread, but there's a lot of unwritten context, so I am sorry if I rehash.  I feel pretty strongly that spec is wrong for this, and all the machinery around locking users and controllers to specific fields under spec reinforces that feeling.

We use the "reconstructible from observation" criteria for status, though we're not always 100% consistent, and the truth is that status being erased is actually a very small risk.  Is that corner case really worth this complexity?

What is the corner case?  Can we spell it out really clearly?

Consider an abstrct resource.

Time t:
- Node has allocatable=4
- Node has 3 pods, each with request=1, allocated=1, actual=1
- Sum(all allocated) = 3 => OK

Time t+1:
- Pod p1 request is changed to 2
- Sum(other allocated, new request)== 4 => OK
- Kubelet sets allocated to 2

Time t+2:
- Pod status is wiped and kubelet crashes before actuating the change
- Kubelet observes pods p2, p3 with request=1 and their cgroups confirm that
- Kubelet sets those pods to allocated=1, actual=1
- Kubelet observes pod p1 with request= 2 but the cgroup says 1
- Kubelet can infer that this was in-flight and set allocated=2, actual=1

Tim t+3:
- Kubelet actuates p1's resize
- Kubelet sets actual=2

That seems fine.  So is the problem centered around multiple concurrent resizes?  Or am I just missing it?

Reply:
> Time t+2:
> - Pod status is wiped and kubelet crashes before actuating the change
> - **Kubelet observes pods p2, p3 with request=1 and their cgroups confirm that**

That is not possible today for memory request, it is not stored anywhere in the cgroups. I'm not sure if ephemeral-storage has similar discoverability issues as well (not in scope of this KEP but someone asked to support it).

Anyways, in above scenario at T+2, now end up in AddPod without knowledge of what was actually granted to p2, p3 (and p1) in the first place. Their request might have changed in the interim isn't any good.

Apologies I dumped the top of that rather huge discussion thread on you. A cleaner analysis of how this works with a way to store allocated is in the following comment and below: https://github.com/kubernetes/enhancements/pull/686#discussion_r308785033

Reply:
> That is not possible today for memory request, it is not stored anywhere in the cgroups

This is a good point, which I missed.  I'm still not convinced that pushing this back into the user-facing API is the right tradeoff.  I know kubelet doesn't like local checkpoints, but I maintain that it is unavoidable for exactly these sorts of reasons.

That thread basically says "we'll use a local checkpoint", then does all of the analysis for that path, showing that it is the better option, then says "we've changed our mind" and proposes the worse option.  What am I missing?

That said you DO know the `limits` from cgroups and you can re-calculate how you would have reacted to any resize request to arrive at the same decision.

Guaranteed class is handled because request == limit.  So let's look at burstable class.  Request must be >= limit at all times.  So you have to consider:

* if memory.limit_in_bytes == limit, then this pod is done resizing (unless we also tweak other knobs?)
* if memory.limit_in_bytes > limit, then this pod must need to be shrunk
* if memory.limit_in_bytes < limit, then this pod must need to be grown 

Right?

Reply:
In all fairness, local checkpoints with atomic file-writes was my preference and showed up in my analysis on that thread. SIG-node folks were against it from get-go based on their history with Borg, I'm not privy to the details.

And correct on the above burstable case for limits. Missing requests is the issue.

Adding ResourcesAllocated to the Spec served two purposes:
1. Allowed kubelet to handle restarts with a single source-of-truth.
1. Allowed user/VPA to determine (via timeouts) if resize was not accepted (Resources != ResourcesAllocated at timer expiry)
imho, it is quite clean.

Reply:
> Missing requests is the issue.

But is it really an issue?  If I know that limit must be >= request, I can use limit as a stand-in.

We have 2 types of information in this calculus.  Information that we use to program cgroups (limits) and information that we use to reason about feasibility and other things (requests).  The memory limit is trivially re-discoverable.  The memory request might have changed while kubelet was down, but why is that a problem?  Either we have completed that resize in cgroups, or we haven't, the rest is recomputable.

Concretely:

* { request = 10, limit = 10}; accept resize to {request = 20, limit = 20} => if cgroups says 20, it is done.  If cgroups says 10, it is not done.  If cgroups says in-between, it must have been in-flight.  No problem.

* { request = 10, limit = 20}; accept resize to {request = 19, limit = 20} => no change needed (cgroups already says 20)

* { request = 10, limit = 20}; accept resize to {request = 20, limit = 20} => no change needed (cgroups already says 20)

* { request = 10, limit = 20}; accept resize to {request = 20, limit = 30} => if cgroups says 30, it is done.  If cgroups says 20, it is not done.  If cgroups says in-between, it must have been in-flight.  No problem.

In all cases, if we find that cgroups does not reflect the current limit we should reconsider feasibility and decide to proceed with resize, do nothing, or kill the pod.

What am I missing?

> After implementing it, imho, it looks quite clean.

If something is in `pod.spec`, the user should own it.  Period.  How are you proposing admission control decide who is and is not allowed to set `spec.resourcesAllocated`?

Again, I am sorry to be a pain in the butt.  There's a philosophy to the API resources that this is violating, and before I give that up, I want to REALLLLLLLY understand why that is the best tradeoff.

Reply:
> But is it really an issue? If I know that limit must be >= request, I can use limit as a stand-in (especially since you CAN actually use this from status except in the miniscule chance that status goes away).

Yes it is - and for starters, you may not have always limits information. Requests-only burstable pos are valid.
Assuming status is lost, however small the chance, consider this going with requests alone:

Node allocatable (requests)=20 with burstable pods p1 {req=10, no limits} , p2 {req=5, no limits}, Total: req=15.
1. Kubelet restarts, nothing changed while offline. We add p1 at req=10, p2 at req=5, everything is good.
2. Kubelet stops, p1 {req=20, no limits}, kubelet back online. We add p1 at req=20, p2 rejected - out-of-resources.

Now, in your scenario even if cgroup limits are available and unchanged during kubelet restart, but requests have increased, we wouldn't know which pods were granted what it is (now) requesting, or how much was granted prior to restart.

Our first prototype was based on information stored in the status as we didn't know it could be lost. All our design choices that relied on any non-reconstructible/state information was shot down, and the story with PodPhase clearly explained why.

---

> If something is in pod.spec, the user should own it. Period. How are you proposing admission control decide who is and is not allowed to set spec.resourcesAllocated?

@liggitt suggested looking into admission control as alternative  for [subresrouce to set ResourcesAllocated that we proposed](https://youtu.be/faRARV3C7Fk?list=PL69nYSiGNLP0OWp38tPBc-jSlMmwWr6Ci&t=2978) for API review.

We initially proposed ResourcesAllocated + pods/resourcesallocated subresource (allowed for kubelet) based on NodeName + /binding precedent we currently have. In that case, if user sets NodeName, scheduler ignores it. If not, scheduler uses the Binding endpoint to set it, and user cannot change it.

Admission plugin idea is better because it allows us to handle down-level clients in addition to setting defaults and access control for ResourcesAllocated and ResizePolicy fields. Also, noderestriction plugin does similar thing where it controls what a fields of Pod object a node is allowed to set.

Reply:
> Node allocatable (requests)=20 with burstable pods p1 {req=10, no limits} , p2 {req=5, no limits}, Total: req=15.
> 
>  1. Kubelet restarts, nothing changed while offline. We add p1 at req=10, p2 at req=5, everything is good.
>  2. Kubelet stops, p1 {req=20, no limits}, kubelet back online. We add p1 at req=20, p2 rejected - out-of-resources.

Thanks for the example.  I guess I assumed that Kubelet could tell the difference between a pod that was running (i.e. was accepted) and a pod that is not running (may or may not be accepted).  In this case, I think you're right.  You can't tell which pod is errant.  Using status as your "checkpoint" you'd have to define the case of "what happens if status gets wiped".  As minute as that case is, it does sort of fail the sniff test for "reconstructible".

Now, if we had a better cgroup mechanism (e.g. maybe we set soft_limit from requests), then that becomes plausible again.  Actually, no.  Because the node could reboot and come back quickly, and we would again lose that information.

I understand the motivation better, but I still feel like this is a local problem, and jamming it into the API is clunky.  I don't understand the desire to not checkpoint locally.  Let me see if I can understand it better.

> @liggitt suggested looking into admission control as alternative for subresrouce

I really want to understand why, here, too.  A subresource seems strictly better to me.  

> We initially proposed ResourcesAllocated + pods/resourcesallocated subresource (allowed for kubelet) based on 
> NodeName + /binding precedent we currently have. In that case, if user sets NodeName, scheduler ignores it. If not, 
> scheduler uses the Binding endpoint to set it, and user cannot change it.

That's not true.  If the user sets NodeName, the scheduler respects it.  By analogy, a sub-resource for `resources` is exactly correct :)

> Admission plugin idea is better because it allows us to handle down-level clients in addition to setting defaults and access control for ResourcesAllocated and ResizePolicy fields.

What criteria decides who is allowed to write these fields and who is not?

> Also, noderestriction plugin does similar thing where it controls what a fields of Pod object a node is allowed to set.

I don't think that's helping the case.  This *also* should have been a subresource, probably.

Reply:
@liggitt Could you please advise if we could go with subresource to set resourcesAllocated field (our original plan when we did the Live API review last Nov in San Diego)? Jordan wanted to see if there were other alternatives [see review session video time 0:56:45](https://youtu.be/faRARV3C7Fk?t=3404) , and suggested I look into admission plugins. 

IIRC, after this review session at contributor summit, the 3 of us talked about this and I came away with the impressions that we were starting to add a subresource - that it was "an imperative operation in an otherwise declarative flow". 

>That's not true. If the user sets NodeName, the scheduler respects it. By analogy, a sub-resource for resources is exactly correct :) 

I agree, that's the precedent we saw as well and proposed subresource for API review - I believe that's the version of the KEP you read in Dec 2019 and LGTM'd it ... before I updated the KEP (after prototyping admission plugin and saw it was a good working alternative to limit access to resourcesAllocated instead of subresource). 

>What criteria decides who is allowed to write these fields and who is not?

podresourceallocation plugin checks [user identity](https://github.com/kubernetes/kubernetes/pull/92127/files#diff-ea00831303a4022109d69aeaef24f074R189) to see if it is a node, and that the node owns the pod whose resourcesAllocated it is trying to write to. The [same check is used](https://github.com/kubernetes/kubernetes/blob/release-1.18/plugin/pkg/admission/noderestriction/admission.go#L124) in noderestriction plugin.

We have to check with @liggitt and get on the same page - subresource or admission plugin. Even if we went with subresource to set resourcesAllocated, I believe I would still need to keep the admission plugin for handling default-setting and downward clients.

Reply:
I'll ask folks in sig-node about why not local checkpointing and get some details - i don't know the issues they had faced in Borg.

Having ResourcesAllocated in spec has two benefits: 
- It opens the path to what's likely a useful feature - give users the ability to specify minimum requests vs ideal resources (not in scope today - but seems like a low-hanging fruit for a later enhancement)
- positive ACK when resize is accepted, a component like VPA can keep a running avg of expected resourceAllocated update time, and if it goes beyond + grace-period, conclude that resize wasn't possible and evict to resize on another node if allowed by its policy config.

So really it's not so much 'shift a problem to the API'. (I myself was initially in favor of local checkpointing just for simplicity)

Reply:
I see the "potential feature" as very different from what Allocated means today.

I got a chance to catch up with @liggitt today, and I don't think I am overstating to say he agrees that this is "weird" in spec.

So I offer this "alternate proposal".  It's a smallish change from the API POV, but makes the spec/status split "regular" and the same as everything else in the system.  I'm not 100% sure we NEED ResourcesAllocated if we also add Conditions, but I am not 100% sure we don't need it either.  And I don't object to it.

Reply:
Sig-node has had problems in the past with local checkpoints, and avoids them when possible.  Kubernetes controllers--the kubelet included--shoudn't be stateful.  But we have broken that design principle in the past, and I understand if that point alone isn't a convincing reason to put something in the pod spec.

One of the previous iterations of the in-place updates proposal had the scheduler involved in resize admission decisions.  We removed that to simplify the proposal, but I think there is a strong desire to do that eventually.  The current design would enable this switch to happen without any additional API changes.

More generally, the split in spec of Resources into Resources and ResourcesAllocated, which correspond to "what I would prefer to run with" and "what I must have to run" are useful primitives for constructing an auto-resizing system.  It enables any actor with the required information to make the decision that it safe to set ResourcesAllocated = Resources.  I'll admit that I don't expect users to set ResourcesAllocated directly--at least not often.  But just like setting Spec.NodeName allows controllers to make intermediary decisions to more completely specify a pod, updating ResourcesAllocated is a further specification of how the pod should be run.  I agree that following the late initialization pattern for a field which can be updated more than once is unprecedented, but that seems logical given that we are allowing scheduling parameters to change post-scheduling for the first time.

I agree that having ResourcesAllocated in the spec smells: We have fields in the spec which we don't _really_ intend for  users to set.  At the same time, I think having it in the status as an "observation" of a local file we wrote smells too.  It isn't _really_ an observation of the current state of a pod on the node.  From the kubelet's perspective, it is the state which it is trying to reconcile pods towards, not an observation of pods.

Reply:
@thockin An earlier iteration of design relied on PodConditions and it was stateful relied on Status not being wiped, and faced problems similar to PodPhase [as David explained](https://github.com/kubernetes/enhancements/pull/686#discussion_r264914522). 

Let's look at the alternate proposal - say we put ResourcesAllocated in Status and do local checkpoint. Consider this possible flow:
1. T=0: User posts pod P1 to apiserver, requests.cpu=1
1. T=1: Scheduler sees unbound Pod, finds it node N1, and invokes /binding
    - Note that we need to extend bind endpoint to set PodStatus.ContainerStatuses[x].ResourcesAllocated=1cpu in addition to NodeName as atomic unit of data.
1. T=2: VPA patches spec (or calls subresource) to set requests.cpu=2
1. T=3: Status gets wiped.
1. T=4: Kubelet sees P1, doesn't have a pod record in local checkpoint, so cannot add the pod as requests.cpu value cannot be trusted to be what scheduler based its decision on.

You might have noticed in step 2 above, we are effectively storing non-recoverable state information in Status field for kubelet to consume in step 5, violates API conventions I think. Since node needs to consume this field to setup the pod, we can argue it really belongs in Spec, and not status.

Potential feature aside, in current design, users **can** set ResourcesAllocated but only on create - which congruent with NodeName precedent, and NodeName has been working well for far, I think.

I agree we don't have a precedent for a system-managed field (like NodeName) being write-more-than-once, we just may not have needed to - until now. This is new territory and we should certainly look very closely and assess risks.

Reply:
>  "what I would prefer to run with"

This is a very different datum than "what I have been allocated".  I'm not
out-of-hand rejecting this idea, I am saying it's DIFFERENT than
`resourcesAllocated` as this doc describes

> I'll admit that I don't expect users to set ResourcesAllocated directly--at least not often. But just like setting Spec.NodeName allows controllers to make intermediary decisions to more completely specify a pod, updating ResourcesAllocated is a further specification of how the pod should be run

If the field is mutable, you have to be ready for it to change.  Spec
represents the desired result.  `resourcesAllocated` is not the desired result.
It is state.

> We have fields in the spec which we don't really intend for users to set

Do we?  Which ones?

> At the same time, I think having it in the status as an "observation" of a local file we wrote smells too

I am not disagreeing.  For better or for worse, we do not distinguish "user-set
values" from "controller-set values" or from "defaulted values".  Part of the
smell is using something OTHER THAN an RBAC expression to govern access.

> From the kubelet's perspective, it is the state which it is trying to reconcile pods towards, not an observation of pods.

I agree!  So let's drop it from the API altogether.  Now it's CLEARLY a kubelet
problem.  I'm being a bit facetious, but you're actually emphasizing my point.
If the goal is to provide a signal about the status of a resize, let's do that.
Allocated is a checkpointed datum.

Concretely, what if the API had just `spec.resources` and `status.resources`.
When a user or controller writes `spec.resources`, kubelet will react by
either:
* setting a Condition that says "resize in progress" (obvious meaning)
* setting `status.resources` = `spec.resources` (meaning resize is complete)
* setting a Condition that says "resize is deferred" (meaning it is hypothetically feasible  but not right now; will try again later)
* setting a Condition that says "resize is rejected" (meaning it will never be feasible)

Given that, why do we need `resourcesAllocated` any more?  Kubelet still needs
it internally, but it's not part of the API.

> Potential feature aside, in current design, users can set ResourcesAllocated but only on create - which congruent with NodeName precedent, and NodeName has been working well for far, I think.

But `nodeName` is NEVER mutable by anyone after create, which is substantially
different than this.

Reply:
> Concretely, what if the API had just `spec.resources` and `status.resources`.
> When a user or controller writes `spec.resources`, kubelet will react by
> either:
> 
> * setting a Condition that says "resize in progress" (obvious meaning)
> * setting `status.resources` = `spec.resources` (meaning resize is complete)
> * setting a Condition that says "resize is deferred" (meaning it is hypothetically feasible  but not right now; will try again later)
> * setting a Condition that says "resize is rejected" (meaning it will never be feasible)
> 
> Given that, why do we need `resourcesAllocated` any more? Kubelet still needs
> it internally, but it's not part of the API.
> 

At one point, we had [an iteration of this KEP](https://github.com/kubernetes/enhancements/pull/686/files/574737cedcdc58e5a4097c8a8e1e50a787b25a4a#diff-57e345a188df5653422eb0ed20bf9442R239) similar to what you suggest, with Status.ResourcesAllocated + PodConditions - that was before we realized Status could be lost and we need a reliable source of truth for ResourcesAllocated.

With current design, scheduler only needs to look in its pod cache at Spec.ResourcesAllocated for fit decision and node resource use computations, really simple.

With that previous design, scheduler had to rely on max(Spec.Resources.Requests, Status.ResourcesAllocated) as Status.ResourcesAllocated & podConditions aren't available until node starts up the pod and writes the status, and if Status is lost, scheduler wasn't able to recover from it.

> But `nodeName` is NEVER mutable by anyone after create, which is substantially
> different than this.

What is the downside of allowing a field to become settable (in a controlled manner) after create?

Reply:
Sorry for the late response.  Getting back from vacation...

>> "what I would prefer to run with"

> This is a very different datum than "what I have been allocated". I'm not
out-of-hand rejecting this idea, I am saying it's DIFFERENT than
resourcesAllocated as this doc describes

Sorry for the mix-up.  resourcesAllocated is "what I must have to run", or "run with this or die".  resources is "what I would prefer to run with".  The process of updating "run with this or die" based on "what I would prefer to run with" is easier to think of as "admitting" the resource change.  "Admitted" pods count against "allocatable" resources, which is probably why the naming ended up as it did.

> If the field is mutable, you have to be ready for it to change. Spec
represents the desired result. resourcesAllocated is not the desired result.
It is state.

I agree.  I'm personally in favor of the field being settable or updatable at any point by any user with access to spec.

resourcesAllocated is the desired result.  There is literally a loop in the kubelet that looks at resourcesAllocated and applies it to existing containers.

[**skip this section if you want**] I was philosophizing about this over my long weekend...   I actually think the most "kubernetes-style" way to handle this sort of intermediate desired state is to have hierarchies of objects (like deployment vs replicaset) in which changes to the upper layer (deployment) can be processed/validated by an intermediary controller before being propagated to the lower layer object (replicaset).  The Spec of the replicaset _is_ the desired result for the replicaset, even though it isn't necessarily the desired result of the deployment.  In theory, we could do away with the late initialization pattern all-together with this approach.  We could have had UnscheduledPod (without pod.nodeName) and Pod (with pod.nodeName) and eliminated the need for binding.  The scheduler would just create fully specified Pod objects based on UnscheduledPod objects once it has made a scheduling decision.  I imagine we decided against this for performance reasons...

> You still have to, don't you? If I submit a resize and kubelet takes 2 seconds to update the allocated field, I don't want to start jamming more pods in there and overflow it. It seems to me that an unambiguous signal is best. Kubelet has either said what is happening or will soon say what is happening. No timeouts, no inferences.

You don't have to look at status.Resources, just spec.Resources and spec.ResourcesAllocated.  This is because changes to container requests are generally accepted (unlike changes to limits, which may not be possible due to current usage).  If in the future the scheduler writes the ResourcesAllocated field, it only needs to look at ResourcesAllocated to do scheduling.

> Is there philosophical or technical objection to the idea of using Conditions to indicate "in progress", "deferred", and "infeasible"?

I'm generally against conditions which describe equality of two separately-controlled "things", and prefer separate observations of those "things".  Assuming, for example, `"in progress" == false` implies `spec.Resources == status.Resources`, as soon as spec.Resources is updated by the user or VPA, the condition becomes incorrect.  You could try and use probe and transition times for the condition.  But even then, there can be races between the kubelet and writers.  For example:

1. kubelet reads spec.CPUs == 1
2. Writer writes spec.CPUs == 2
3. kubelet observes spec.CPUs == 1
4. kubelet writes status.conditions["in progress"] == false

Even with probe timestamps, in the above case the probe timestamp would occur after the update timestamp.

Conditions also don't do a good job of describing multiple in-progress updates.  If I update CPUs 1 -> 2, and then 2 -> 3, the conditions only apply to the last update I made.  CPUs 1 -> 2 may be accepted by the kubelet, and 2 -> 3 may be deferred, but I can't tell that 1 -> 2 has been accepted.  With an explicit description of what the kubelet has "admitted", I can verify that an explicit update (1 -> 2) has been accepted.

Also, what is the difference between deferred and infeasible?  If deferred implies "will retry" and infeasible means "won't retry", then we would only ever use deferred.

Reply:
> resourcesAllocated is the desired result. There is literally a loop in the kubelet that looks at resourcesAllocated and applies it to existing containers.

We're kind of arguing semantics.  It's KUBELET's desired result.  Pod.spec is not about what kubelet wants, it is about what the user wants.  The user sets their intentions in `resources`, right?


> [skip this section if you want] I was philosophizing

You know me better than to expect I would skip philosophizing :)

> about this over my long weekend... I actually think the most "kubernetes-style" way to handle this sort of intermediate desired state is to have hierarchies of objects (like deployment vs replicaset) in which changes to the upper layer (deployment) can be processed/validated by an intermediary controller before being propagated to the lower layer object (replicaset). The Spec of the replicaset is the desired result for the replicaset, even though it isn't necessarily the desired result of the deployment. In theory, we could do away with the late initialization pattern all-together with this approach. We could have had UnscheduledPod (without pod.nodeName) and Pod (with pod.nodeName) and eliminated the need for binding. The scheduler would just create fully specified Pod objects based on UnscheduledPod objects once it has made a scheduling decision. I imagine we decided against this for performance reasons...

There was once a BoundPod resource, which isn't quite the same but gets at the same result.  That was a monolithic thing - image one object with 100 podspecs in it.  Yikes!

I'm not sure such a model as you propose was considered, frankly, but whether it was or wasn't, it's not where we landed.

That said, it may be worth considering for this.  What if we had a PodResize "pseudo-imperative" resource?  I am not sure it solves any of the problems we have here.  We still have a mutable `resources` field in pod which suffers the same RBAC weirdness.

>>> With that previous design, scheduler had to rely on max(Spec.Resources.Requests, Status.ResourcesAllocated)
>>
>> You still have to, don't you? If I submit a resize and kubelet takes 2 seconds to update the allocated field, I don't want to start jamming more pods in there and overflow it. It seems to me that an unambiguous signal is best. Kubelet has either said what is happening or will soon say what is happening. No timeouts, no inferences.
> 
> You don't have to look at status.Resources, just spec.Resources and spec.ResourcesAllocated. This is because changes to container requests are generally accepted (unlike changes to limits, which may not be possible due to current usage). If in the future the scheduler writes the ResourcesAllocated field, it only needs to look at ResourcesAllocated to do scheduling.

My point was that you still need to take a max().  No one field is sufficient, whether it is from spec or status isn't really important.

Given a spec (desired) and status (current) and condition (opinion about what spec-status disparity means), I think you can still make the same decisions.

If status is blank, you can't assume anything about the pod and you need to wait for a status update.

If spec == status, no confusion

If spec != status && condition(ResizeInProgress) - you know you can take the max().

If spec != status && conditon(ResizeInfeasible) - you know you can just use status and ignore spec.

If spec != status a&& condition(ResizeDeferred) - you know the resize might be feasible, but what that means depends on context.  The scheduler might "score" that node lower if it would depend on deferred-resizes.

>> Is there philosophical or technical objection to the idea of using Conditions to indicate "in progress", "deferred", and "infeasible"?
>
> I'm generally against conditions which describe equality of two separately-controlled "things", and prefer separate observations of those "things". Assuming, for example, "in progress" == false implies spec.Resources == status.Resources, as soon as spec.Resources is updated by the user or VPA, the condition becomes incorrect. You could try and use probe and transition times for the condition. But even then, there can be races between the kubelet and writers. For example:
> 
> kubelet reads spec.CPUs == 1
> Writer writes spec.CPUs == 2
> kubelet observes spec.CPUs == 1
> kubelet writes status.conditions["in progress"] == false
> Even with probe timestamps, in the above case the probe timestamp would occur after the update timestamp.
> 
> Conditions also don't do a good job of describing multiple in-progress updates. If I update CPUs 1 -> 2, and then 2 -> 3, the conditions only apply to the last update I made. CPUs 1 -> 2 may be accepted by the kubelet, and 2 -> 3 may be deferred, but I can't tell that 1 -> 2 has been accepted. With an explicit description of what the kubelet has "admitted", I can verify that an explicit update (1 -> 2) has been accepted.

Excellent points.  Conditions are a poor fit for what I am describing here, but not only for the reasons you cite.  Conditions really want to be orthogonal, and these are not.  If we pursue this model, it may be better to actually enumerate this as a distinct think in the API.

As for races, there are always races in our eventually-consistent model.  We get away with it by defining the edge-cases and making sure that losing the race will still self-recover.

In the case you lay out, consumers could observe that `spec.resources` != `status.resources` AND the resize indicator said no resize was in progress, and it was not deferred, and it iwas not infeasible.  So clearly it hasn't been updated.

But there's STILL a race - if there was another resize in progress, a client could observe that condition and assume it applies to the wrong thing.  I was trying to stay very high-level, but maybe we'd actually want a map of resource name to resize-status which is automatically cleared on any update to the corresponding `spec.resources`.

So in the timeline you listed:

kubelet reads spec.CPUs == 1
user writes spec.CPUs = 2
  apiserver sets status.resize["cpu"] = Unknown
kubelet tries to set status.resize["cpu"] = 1 (why?)
resourceVersion blocks kubelet from writing
kubelet observes change to status.resize["cpu"] and must decide what to do
kubelet sets status.resize["cpu"] to one of (pending, deferred, infeasible)

Or maybe we use `metadata.generation` as part of the resize status.  "pending as of 1234" while the current generation is "1235" means there are potentially unacknowledged changes.  But this is very coarse, so I don't like it

Note that through all this - the API is only carrying either user-intention or re-observable state.  If you need to keep track of kubelet's allocated field beyond this, that's still "internal".

> Also, what is the difference between deferred and infeasible? If deferred implies "will retry" and infeasible means "won't retry", then we would only ever use deferred.

I assumed (maybe wrongly) that Deferred means something like "the CPUs you asked for exist on the machine but are in use, so I will try again the next time a Pod is deleted", whereas Infeasible means something like "You asked for 17 cpus on a 16 core node, this will never work and I won't bother to re-evaluate it".  Does that make any sense?

Reply:
David, your reply broke the thread, so I am trying to glue it back into place.  I will leave the whole thing quoted.


>> The user sets their intentions in resources, right?
>
> That is the common case for sure. I'm not arguing users will often set resources allocated, or even should generally set it. I'm only arguing that
>
> The field has clearly defined behavior, such that users could set it.
>
> a. To reiterate, setting resourcesAllocated means "run with these resources, or don't run at all". This is as opposed to spec.resources, which is "what I would prefer to run with". Setting both results in "start with resourcesAllocated, but then try and change to resources".
> The field is not linked to an observation of the state of any "thing" that makes up the pod (containers, sandboxes, images, etc).
> The field is generally used as a description of what should be. Updating the field results in "the system" making changes in response.

Hypothetically, if a user changed this, Kubelet would have to immediately kill the pod if it could not start the actuation process *right now*, right? 

> I can understand putting this information in Status, even if I don't like it.

I appreciate that.  If we loosened the restrictions on who can write to the field, as described above, I could possibly be convinced that `spec` is acceptable (though the name is not great).  The root of the problem I have with it is the unprecedented access control.

If the semantics were "regular", the discussion would then be more about whether this is a useful feature or not.  Is it?  It feels marginal to me - do we have concrete cases where expressing "I want X, take your time" vs "I need X NOW, or else kill me" is critical?  Specifically, let's assume that kubelet had an unambiguous status signal of "this resize has been considered and deferred" in pod, so a controller doesn't need to time out.

Given "I want X, take your time", and that status signal, I can choose to do the kill from outside.  Devil's advocate: If I am doing a resize, I expect that MOST of the time, I want it ASAP, and if I can't have it on node A, move me to node B.

> We don't appear to have been consistent on such fields. Spec.NodeName is the closest precedent IMO, but there are other fields, such as spec.Priority (the integer, not the name), which are not generally expected to be set by users (expected to be set by admission controllers). On the other hand, something like status.nominatedNodeName, or even status.QoS could be argued to meet the above points.

NodeName is settable by users.  That's a niche "feature", but it is "regular".  If this were regular (no bizarre access control), I'd hate it a lot less.

>> I was trying to stay very high-level, but maybe we'd actually want a map of resource name to resize-status which is automatically cleared on any update to the corresponding spec.resources.
>
> Keep in mind that there can be multiple resizes in-progress on a single resource.

For a given pod there's only "where am I now" and "where am I going", right?  If I resize memory from 4GB to 1GB (which might take time) and then change it to 3GB, there's no guarantee that the 1GB size will ever be realized, if I understand correctly?

> I believe the problem of tracking a single update is solved decently by the deployment with status.ObservedGeneration, which I think is what you are suggesting above. But it does then make it harder for users to reason about.
>
> Given spec.Resources and status.Resources, we can already determine if a resize is "in progress",

How so?  What if Kubelet has not seen the resize?  Or what if it has decided it is not feasible roght now, but when the next pod exits, it will be feasible?  Or what if you asked for more than the system can ever give you?

When I say "in progress" I mean that we are actively actuating it,

>so given all the hoops a user has to jump through to correctly make use of it (assuming we implemented ObservedGeneration or something else), I think it is less useful than the fields we already have. The same goes for "Infeasible" which can be discovered by looking at node.status.Allocatable and pod.spec.Resources. FWIW, I haven't seen anyone ask for an "infeasible" status either.

You're right that infeasible can probably be determined by other means.  But we still have to either prevent such a resize from happening (e.g. validate that resources <= allocatable in apiserver) or we have to handle it.

Without a clear signal, all we can do is time-out.  That's not great, IMO.

> The two main unanswered questions (given just spec.resources and status.resources) are:
>
> 1. why hasn't my update been applied?
> 2. what portion of the node is "occupied" by a given pod?
>
> The current possible answers to 1 are either "because there isn't room for it" or "because the updates haven't been applied yet". We can either answer this with a "here's what there is room for" (~status.resourcesAllocated) or an A or B condition (~"deferred") + observed generation.
>
> Answering 2 is harder to answer because "occupied" is really about what set of resources would pass the scheduler's scheduling algorithm, and the kubelet's admission algorithm. Today, this question is easy to answer, because the scheduler and kubelet both use the same definition of occupied: spec.resources.requests. But if now kubelet is using (internal) resourcesAllocated checkpoint, and scheduler is using max(spec.resources, status.resources), we can get different answers for resizes vs creates. Creates go through the scheduler, and thus "occupied" == max(spec, status). Resizes only involve the kubelet, so "occupied" == (internal) resources allocated. For something like kubectl top no, we would probably display it just using spec.resources, which would make nodes appear overcommitted by requests.
>
> The broader point is that without resourcesAllocated somewhere in the API, we lose our sum(pod resources) < node allocatable invariant. It technically is still maintained in a file on the node, but I can't use it for chargeback, or for determining occupancy of a node.

Depending on what context, you probably wnat different views of this anyway?  For chargeback, I'd argue spec.resources is correct.  For occupancy, assuming the max() is safe, but pessimistic, so any status about "allocated" that we offer becomes an optimization.

> TBH, doing neither, and just having events also seems like a reasonable position for alpha. My preference is roughly (spec.ResourcesAllocated > nothing/punt until beta > status.ResourcesAllocated > status.conditions["deferred"]).

I'm going to s/conditions/signal/ because (as you called before) literal conditions are a poor fit.  I think signal serves two purpose - one to patch holes in a design that lacks "allocated" and the other to actually indicate to users what is going on without a timeout.  The former is obviously only useful if the design lacks allocated, but the latter (I think) is important either way.  IOW, even if we keep spec.allocated, I think signals should be considered strongly.

I'm OK with punting unknowns, but we will have to come back to them, even if just to rule them out.  

Instead of recapping possible designs, I'd like to hear your thoughts on regularizing the spec.allocated: get rid of the access control, define the full gamut of relationships between `resources` and `resourcesAllocated`, and maybe think abut a more intent-oriented (rather than state-oriented) name.  Then it would follow `nodeName` much more accurately.

After that we can discuss signals as a distict topic.  If we can't regularize that, they get mixed in with the rest.

Reply:
> Hypothetically, if a user changed this, Kubelet would have to immediately kill the pod if it could not start the actuation process right now, right?

Yes.

> Devil's advocate: If I am doing a resize, I expect that MOST of the time, I want it ASAP, and if I can't have it on node A, move me to node B.

I actually think this is a reasonable position.

When updating a pod's spec.Resources, you would have to assume the pod will be killed for, e.g. PDB considerations, which may be hard.  It certainly doesn't allow VPA to provide any policy knobs, such as grace periods.  It is just a one-resize-fits-all approach.  I imagine it would be also be useful to kick an unimportant pod off of the node instead of moving the pod being resized.  The scheduler could do that only if it actually has a chance to move pods before the kubelet nukes the pod being resized.  I believe the motivation behind the feature is to be able to reduce the number of restarts for pods that don't tolerate restarts well.  Leaving open the possibility for "something else" to come in and make room for the in-memory database to expand its memory footprint seems useful, even if we haven't built the "something else" today.  IIUC, we do want to support "try to resize, but don't kill" policies.

> When I say "in progress" I mean that we are actively actuating it,

Got it; i misunderstood.  We can set it to true when the kubelet process is scheduled by the kernel, and back to false when it is removed, just to be safe ;).  IMO, we should think about this in terms of how one would build an SLO around this.  Controllers may be unable to perform the resize for a variety of reasons, not limited to it being deferred, such as container runtime unavailability.  We want to build a system in which a user can articulate "when my workload needs additional resources, provide them to my pod within 2 minutes 99% of the time".  I think events can supplement this to provide the "why", including which controllers attempted, but failed.

> Instead of recapping possible designs, I'd like to hear your thoughts on regularizing the spec.allocated: get rid of the access control, define the full gamut of relationships between resources and resourcesAllocated, and maybe think abut a more intent-oriented (rather than state-oriented) name. Then it would follow nodeName much more accurately.

I like the PreferredDuringSchedulingIgnoredDuringExecution vs RequiredDuring... wording, so I'm going to borrow it, and use a Preferred/Required to differentiate the two sets of resources.

I'm going to do something crazy, and invert what we have been naming things, just to confuse you. ;).  I think the naming is slightly better if we use `Resources` to mean essentially what it does today, which is the resources that a running pod has.  Resources is now mutable.  If you change it, you either:

a. update the resources of your container.
b. are killed, if there isn't room for the update.

`PreferredResources` is a new field that means "id prefer to eventually have these resource requests, but don't want to be killed to get it".  "Something" in kubernetes will compare `PreferredResources` with `Resources`, and will update `Resources` if it has determined it can be done without resulting in the termination of your pod.  Or, it could choose to be "fancy" and find a way to make room for the additional resources you want.

I could (either by creating a new pod, or by updating an existing pod), have `PreferredResources` > `Resources`, in which case "something" will check to see if it is safe to change `Resources` to match, otherwise will retry-backoff.  I could also have `PreferredResources` < `Resources`, which would do the same (although downward changes to Resources should always succeed).

If `PreferredResources` is not set (e.g. because i'm using an older client), then `Resources` is still a statement of fact, and no controller will try to update it to match.

One oddity is that if `PreferredResources` == `Resources`, but then I lower `Resources` without changing `PreferredResources`, "something" would change it right back.  I suppose one contrived use-case would be to temporarily reduce the amount of resources assigned to less important pods (e.g. to influence out-of-eviction handling in-favor of a pod that needs more right now), with the intention that the less important pod either be granted the additional resources at a later point, or moved to another node.

Reply:
>> Hypothetically, if a user changed this, Kubelet would have to immediately kill the pod if it could not start the actuation process right now, right?
>
> Yes.
>
>> Devil's advocate: If I am doing a resize, I expect that MOST of the time, I want it ASAP, and if I can't have it on node A, move me to node B.
>
> I actually think this is a reasonable position.
>
> When updating a pod's spec.Resources, you would have to assume the pod will be killed for, e.g. PDB considerations, which may be hard. It certainly doesn't allow VPA to provide any policy knobs, such as grace periods. It is just a one-resize-fits-all approach. I imagine it would be also be useful to kick an unimportant pod off of the node instead of moving the pod being resized. The scheduler could do that only if it actually has a chance to move pods before the kubelet nukes the pod being resized. I believe the motivation behind the feature is to be able to reduce the number of restarts for pods that don't tolerate restarts well. Leaving open the possibility for "something else" to come in and make room for the in-memory database to expand its memory footprint seems useful, even if we haven't built the "something else" today. IIUC, we do want to support "try to resize, but don't kill" policies.

These are great points, and I think they are not mutually exclusive with what users would "naively" want.  I want it ASAP.  If you can do it in 30 seconds by evicting something, that may well be faster than killing me and moving me.`

>> When I say "in progress" I mean that we are actively actuating it,
>
> Got it; i misunderstood. We can set it to true when the kubelet process is scheduled by the kernel, and back to false when it is removed, just to be safe ;). IMO, we should think about this in terms of how one would build an SLO around this. Controllers may be unable to perform the resize for a variety of reasons, not limited to it being deferred, such as container runtime unavailability. We want to build a system in which a user can articulate "when my workload needs additional resources, provide them to my pod within 2 minutes 99% of the time". I think events can supplement this to provide the "why", including which controllers attempted, but failed.

SLOs are going to be great at an aggregate level, but when it is my pod that isn't resized yet, I just want to know why.  That's what I keep coming back to with signals in general.  Lack of signal is a poor signal.  "Can you hear me?  Hello?  Is this thing on?"

> Instead of recapping possible designs, I'd like to hear your thoughts on regularizing the spec.allocated: get rid of the access control, define the full gamut of relationships between resources and resourcesAllocated, and maybe think abut a more intent-oriented (rather than state-oriented) name. Then it would follow nodeName much more accurately.
>
> I like the PreferredDuringSchedulingIgnoredDuringExecution vs RequiredDuring... wording, so I'm going to borrow it, and use Preferred/Required to differentiate the two sets of resources.
>
> I'm going to do something crazy, and invert what we have been naming things, just to confuse you. ;). I think the naming is slightly better if we use Resources to mean essentially what it does today, which is the resources that a running pod has. Resources is now mutable. If you change it, you either:
>
> a. update the resources of your container.
> b. are killed, if there isn't room for the update.
>
> PreferredResources is a new field that means "id prefer to eventually have these resource requests, but don't want to be killed to get it". "Something" in kubernetes will compare PreferredResources with Resources, and will update Resources if it has determined it can be done without resulting in the termination of your pod. Or, it could choose to be "fancy" and find a way to make room for the additional resources you want.

This feels better.  It has regular meaning.  There's still an issue with access control - does Kubelet ever write back to pod spec today?  To make this possible it would have to, right?  If I grow my preferredResources, "someone" (I presume kubelet) will have to sync resources to  match it.  That's problematic, but if it is just kubelet maybe not the end of the world.

Does this really solve the checkpoint issue?  I don't think it does.  If I grow `preferred`, and kubelet agrees with that but needs time to actuate, it can't immediately set `resources` because that means NOW.  So you'd still need a local checkpoint to keep track of the fact that it was accepted but not actuated.  I think?

> I could (either by creating a new pod, or by updating an existing pod), have PreferredResources > Resources, in which case "something" will check to see if it is safe to change Resources to match, otherwise will retry-backoff. I could also have PreferredResources < Resources, which would do the same (although downward changes to Resources should always succeed).

Downward changes can fail, can't they?  e.g. you have more stuff in tmpfs than the pod cgroup has memory - the kernel can fail cgroup ops.

> If PreferredResources is not set (e.g. because i'm using an older client), then Resources is still a statement of fact, and no controller will try to update it to match.
>
> One oddity is that if PreferredResources == Resources, but then I lower Resources without changing PreferredResources, "something" would change it right back. I suppose one contrived use-case would be to temporarily reduce the amount of resources assigned to less important pods (e.g. to influence out-of-eviction handling in-favor of a pod that needs more right now), with the intention that the less important pod either be granted the additional resources at a later point, or moved to another node.

It seems like there's also possible conflicts between a replicaset and pods?  If I change the rs to have resources != preferred, would RS controller try to reconcile back to the RS state?

I'm going to chew on this a bit more, but I am a bit skeptical of it still.  It feels complex for little gain.

If we instead assumed that `spec.resources` is the request, `status.resources` is the actual, and we track resize status it seems to express what we really need to express without anything extra.

Reply:
> There's still an issue with access control - does Kubelet ever write back to pod spec today? To make this possible it would have to, right? If I grow my preferredResources, "someone" (I presume kubelet) will have to sync resources to match it. That's problematic, but if it is just kubelet maybe not the end of the world.

I'm being purposefully vague with this, because I don't think it should matter for the API.  I still personally believe the scheduler should be the "someone" performing this process.  It also doesn't preclude users or other processes from doing it.  I hate beating the NodeName drum, but it has similar properties to that field in this case.  The scheduler is generally the place where allocation of compute resources is performed, but you can go around it if you'd like and put something directly on a node.

That being said, we compromised with sig-node leads and agreed to keep everything local to the kubelet, at least to start.  I'm happy to revisit that, of course.

> Does this really solve the checkpoint issue? I don't think it does. If I grow preferred, and kubelet agrees with that but needs time to actuate, it can't immediately set resources because that means NOW. So you'd still need a local checkpoint to keep track of the fact that it was accepted but not actuated. I think?

It is the exact same as the current proposal, with names changed, and fields moved.  If the kubelet agrees, it means it can immediately set spec.Resources.  These are still part of spec, and are thus desired state.  Its OK if we can't immediately actuate.  The main concern is ensuring sum(requests) < allocatable.

> Downward changes can fail, can't they? e.g. you have more stuff in tmpfs than the pod cgroup has memory - the kernel can fail cgroup ops.

Yes, but that is only for limits.  Downward changes for requests can't fail.

> It seems like there's also possible conflicts between a replicaset and pods? If I change the rs to have resources != preferred, would RS controller try to reconcile back to the RS state?

I have to think about this...

Reply:
>> There's still an issue with access control - does Kubelet ever write back to pod spec today? To make this possible it would have to, right? If I grow my preferredResources, "someone" (I presume kubelet) will have to sync resources to match it. That's problematic, but if it is just kubelet maybe not the end of the world.
>
> I'm being purposefully vague with this, because I don't think it should matter for the API. I still personally believe the scheduler should be the "someone" performing this process. It also doesn't preclude users or other processes from doing it. I hate beating the NodeName drum, but it has similar properties to that field in this case. The scheduler is generally the place where allocation of compute resources is performed, but you can go around it if you'd like and put something directly on a node.

I think it matters because granting write-access to resources is a Big Deal.
If we're going to have an API that requires someone OTHER THAN the user write
to pod spec, I think we need to _KNOW_ that it is an important change to make.

Scheduler can't write to pod spec - it can only use pods/binding to set
NodeName. I'm still going to argue that `resize` should be a subresource, so
that VPA has very granular control, so maybe that can cover both spec.resources
and spec.preferredResources together?  How would the flow work?

Success:
- VPA sets `spec.preferredResources` via /resize
- scheduler decides that is feasible and sets `resources`
- kubelet must begin actuation of the resize or kill the pod
- kubelet sets `status.resources` to indicate completion

Fail:
- VPA sets `spec.preferredResources` via /resize
- scheduler decides that is not currently feasible and does not set `resources`
- VPA eventually times it out and makes a new Pod

That's net-new functionality for the scheduler that seems out-of-place to me.
It would seem better to:

- VPA sets `spec.preferredResources` via /resize
- kubelet decides that is feasible and sets `resources` via /resize
- kubelet begins actuation of the resize
- kubelet sets `status.resources` to indicate completion

If the user or VPA sets both `spec.preferredResources` and `spec.resources`,
kubelet must begin actuation of the resize or kill the pod.

>> Does this really solve the checkpoint issue? I don't think it does. If I grow preferred, and kubelet agrees with that but needs time to actuate, it can't immediately set resources because that means NOW. So you'd still need a local checkpoint to keep track of the fact that it was accepted but not actuated. I think?
>
> It is the exact same as the current proposal, with names changed, and fields moved. If the kubelet agrees, it means it can immediately set spec.Resources. These are still part of spec, and are thus desired state. Its OK if we can't immediately actuate. The main concern is ensuring sum(requests) < allocatable.

I see, and because kubelet set it, there's no (or little) risk of it being
infeasible.

>> It seems like there's also possible conflicts between a replicaset and pods? If I change the rs to have resources != preferred, would RS controller try to reconcile back to the RS state?
>
> I have to think about this...

Maybe sufficient to just say that Deployment and RS should always set them to
the same values (or not set Preferred at all, and default that in pod
creation).

Overall: 
This isn't awful but it still feels like a hypothetical feature as an excuse
to not checkpoint.

I still think that proper status signals are needed BUT that doesn't serve as a
replacement for checkpoint.  The only corner-case is "what if status gets
nuked".  That doesnt ACTUALLY happen, but we use it as a litms test.  Are we
going to let that force us down a path we don't know we want?

If, on the other hand, you assume status doesn't get nuked (or kubelet keeps
the checkpoint) then I think the signals we (I) want ANYWAY are sufficient.

To take the example @vinaykul posted up-thread:

> Node allocatable (requests)=20 with burstable pods p1 {req=10, no limits} , p2 {req=5, no limits}, Total: req=15.
> 
> Kubelet restarts, nothing changed while offline. We add p1 at req=10, p2 at req=5, everything is good.
> Kubelet stops, p1 {req=20, no limits}, kubelet back online. We add p1 at req=20, p2 rejected - out-of-resources.

With proper status signals:

```
 t0: allocatable=20
 t1: add p1 at req=10
 t2: add p2 at req=5
 t3: kubelet writes p1 status.resources=10
 t4: kubelet writes p2 status.resources=5
 t5: kubelet goes down
 t6: resize p1 to req=20
     api sets status.resize = Pending
 t7: kubelet wakes up
 t8: kubelet looks at p1, finds status.resources != spec.resources
     also finds status.resize = Pending
 t9; kubelet adds p1 at req=10
t10: kubelet adds p2 at req=5
t11: kubelet decides that p1 is infeasible at req=20
t12: kubelet sets status.resize = Infeasible
```

Making it even worse, still works

```
 t0: allocatable=20
 t1: add p1 at req=10
 t2: add p2 at req=5
 t3: kubelet writes p1 status.resources=10
 t4: kubelet writes p2 status.resources=5
 t5: kubelet goes down
 t6: resize p1 to req=11
     api sets status.resize = Pending
 t7: kubelet wakes up
 t8: kubelet looks at p1, finds status.resources != spec.resources
     also finds status.resize = Pending
 t9; kubelet adds p1 at req=10
t10: kubelet adds p2 at req=5
t11: kubelet decides that p1 is feasible at req=11
t12: kubelet sets status.resize = InProgress
t13: kubelet goes down
t14: resize p1 to req=12
     api resets status.resize = Pending
t15: kubelet wakes up
t16: kubelet looks at p1, finds status.resources != spec.resources
     also finds status.resize = Pending
t17; kubelet adds p1 at req=10
t18: kubelet adds p2 at req=5
t19: kubelet decides that p1 is feasible at req=12
t20: kubelet sets status.resize = InProgress
t21: kubelet goes down
t22: resize p1 to req=16
     api resets status.resize = Pending
t23: kubelet wakes up
t24: kubelet looks at p1, finds status.resources != spec.resources
     also finds status.resize = Pending
t25; kubelet adds p1 at req=10
t26: kubelet adds p2 at req=5
t27: kubelet decides that p1 is feasible, but deferred at req=16
t28: kubelet sets status.resize = Deferred
t29: kubelet goes down
t30: resize p1 to req=20
     api resets status.resize = Pending
t31: kubelet wakes up
t32: kubelet looks at p1, finds status.resources != spec.resources
     also finds status.resize = Pending
t33; kubelet adds p1 at req=10
t34: kubelet adds p2 at req=5
t35: kubelet decides that p1 is infeasible at req=20
t36: kubelet sets status.resize = Infeasible
```

There's no ambiguity there UNLESS status is wiped.  Which, again, doesn't
happen.  With a local checkpoint, all of this can be recoverable.

Reply:
> I think it matters because granting write-access to resources is a Big Deal.
If we're going to have an API that requires someone OTHER THAN the user write
to pod spec, I think we need to KNOW that it is an important change to make.

Sorry for being unclear.  I agree it is a big deal to grant a controller the ability to write to fields in the pod spec.  I just meant it doesn't matter if it is controller A or controller B.  It is all the same to users.

> Scheduler can't write to pod spec - it can only use pods/binding to set
NodeName. I'm still going to argue that resize should be a subresource, so
that VPA has very granular control, so maybe that can cover both spec.resources
and spec.preferredResources together? How would the flow work?

I'm remembering now why we settled on the design we did...  Having spec.Resources == preferred, and spec.ResourcesAllocated == required makes the access control easier, as VPA/users interact with spec.Resources, and scheduler/kubelet (depending on what we decide) interacts with spec.ResourcesAllocated.

Given that, let's go back to spec.Resources (preferred) and spec.**RequiredResources** (old resources allocated).  There would be a subresource for setting spec.RequiredResources, which kubelet/scheduler would use to update spec.RequiredResources based on spec.Resources.

To be clear, I fully support the use of subresources to give more granular access to fields.

> That's net-new functionality for the scheduler that seems out-of-place to me.

The benefit of having the scheduler determine feasibility of resizes is that it can serialize them with scheduling decisions.  Otherwise, a high rate of resizes makes it likely that scheduled pods will be rejected on arrival.

> Overall:
> This isn't awful but it still feels like a hypothetical feature as an excuse to not checkpoint.

It would be a useful feature to have the scheduler make feasibility decisions.  It was originally one of the core requirements of the KEP before sig-node blocked it unless it was dropped from the design.  That being said, a field doesn't need to be a "feature" to be in spec.  It just needs to describe desired state, and have consistent behavior when set.

I don't buy the "excuse not to checkpoint" argument.  There are many things in the API that could have been checkpointed, but that doesn't mean they should be.  NodeName could have been checkpointed in a PV attached to the scheduler, which the scheduler "observes" periodically, and writes to the pod status.  Or, the priority admission controller could, instead of initializing spec.Priority integer during admission, periodically re-resolve the priority of the priority class, and write it to status.  The question isn't "could this field have been recorded locally in the controller, and written to status?".  The question is "does this field describe desired state" (or what _should_ be), or is this an observation of the thing represented by the object?

> I still think that proper status signals are needed BUT that doesn't serve as a
replacement for checkpoint. The only corner-case is "what if status gets
nuked". That doesnt ACTUALLY happen, but we use it as a litms test. Are we
going to let that force us down a path we don't know we want?

The spec vs status litmus test exists for a reason.  It is so that writing controllers is simple and reliable.  They take desired state from a source of truth, observe the actual state of the thing, take corrective actions, and report the actual state as status.  Your example below illustrates my point nicely.  If status is required to know the desired state (i.e. what _should_ be), the above controller strategy no longer works, as the controller would have to ensure state is written to durable storage _before_ taking corrective action.

"Down a path we don't know we want" is fair.  I do think my position (field in spec + scheduler involvement) is a reasonable path that we _should_ want.  However, if the kubelet is forever the only controller for resizes, i'm less certain.  I can see arguments for efficiency (no extra write, just included in status), and simplicity of implementation (no subresources, although we now need checkpoints).

>  t7: kubelet wakes up
>  t8: kubelet looks at p1, finds status.resources != spec.resources
>     also finds status.resize = Pending
> t9; kubelet adds p1 at req=10
> t10: kubelet adds p2 at req=5

We probably don't want to do admission based on the contents of pod status for many reasons.  For one, pod status often lags considerably behind what it has admitted.  For example, assume node has 1 core:

1. Admit pod A w/ 1000m cpu
2. Update pod A status: status.resources.cpu == 1000m
2. Resize pod A to 500m cpu.  Accepted by the kubelet!
3. Admit pod B w/ 500m cpu
4. Kubelet restarts
5. Pod A admitted w/ status.resources.cpu == 1000m
6. Pod B rejected w/ request == 500m cpu

What we can do, which is what we have discussed previously, is checkpoint the "accepted" or "admitted" decisions the kubelet makes.  That way, admission doesn't rely on pod status, just the local file.

> With a local checkpoint, all of this can be recoverable.

Completely agree.  Checkpointing, in theory, can solve any problem posed by a kubelet restarting.

Reply:
A few points as I catch-up..
- Regarding max(resources, resourcesAllocated) or max(spec, status): It is well intentioned but gives a user some control over how cluster resource is applied. Suppose a node with 4G mem has a guaranteed Pod1 (1G,1G), and evil user starts up a burstable pod P2(1G,-). Now, that user can request 3.1G which will lead to ResizeInfeasable/Deferred condition. Observing that, the user switches back to 1G which would clear that condition, quickly follow it by requesting 3.1G ... and keep toggling to take advantage of durations during which scheduler is not selecting that node. This way it can potentially block off additional 2G with some probability without "paying" for it.
- If we used just resourcesAllocated (whether in spec or status), we do increase the probability of hitting the already existing race condition by (duration it takes for kubelet to patch the admitted resize + scheduler watch seeing it). It is sort of 'pick your poision' situation.
- Yes, spec.resourcesAllocated was a very poor choice.. it should have been resourcesToAllocate as Dawn mentioned in the last meeting..
- ```PreferredResources``` (assuming that it is Requests + Limits) sounds like a clean option, let me chew on this a little. Can we use /resize to update spec.Resources if we go with this model? This would avoid needing to do access control in admission plugins, but wouldn't we still need to do access control somewhere outside RBAC on what fields someone is allowed to write - something similar to NodeRestriction plugin to allow node to write only to Resources field.
- +1 for ```/resize``` subresource that can be extended for other objects.. our original use-case was for long-running jobs being resized.
- Our [first proposal](https://static.sched.com/hosted_files/kccnceu19/31/RestartFreeVerticalScaling.pdf) did have scheduler approving resize and kubelet implementing it, however Derek felt that for a bound pod, resize should strictly be a apiserver-node transaction. (scheduler, in the future, may assist if possible by /evict on lower-priority pods)

Reply:
>> Scheduler can't write to pod spec - it can only use pods/binding to set
>> NodeName. I'm still going to argue that resize should be a subresource, so
>> that VPA has very granular control, so maybe that can cover both spec.resources
>> and spec.preferredResources together? How would the flow work?
> 
> I'm remembering now why we settled on the design we did...  Having spec.Resources == preferred, and spec.ResourcesAllocated == required makes the access control easier, as VPA/users interact with spec.Resources, and scheduler/kubelet (depending on what we decide) interacts with spec.ResourcesAllocated.
> 
> Given that, let's go back to spec.Resources (preferred) and spec.**RequiredResources** (old resources allocated).  There would be a subresource for setting spec.RequiredResources, which kubelet/scheduler would use to update spec.RequiredResources based on spec.Resources.

Just to be clear, this path requires:
- a subresource (e.g. /resize?) for users and VPA-like controllers
- a subresource (e.g. /allocate?) for kubelet and/or scheduler
- clear semantics for what happens when a user sets RequiredResources != Resources

It could maybe be simplified to a single sub-resource to cover both fields, at the cost of expanded power granted.

>> That's net-new functionality for the scheduler that seems out-of-place to me.
> 
> The benefit of having the scheduler determine feasibility of resizes is that it can serialize them with scheduling decisions.  Otherwise, a high rate of resizes makes it likely that scheduled pods will be rejected on arrival.

I think this is a solution looking for a problem.

>> This isn't awful but it still feels like a hypothetical feature as an excuse to not checkpoint.
> 
> It would be a useful feature to have the scheduler make feasibility decisions.  It was originally one of the core requirements of the KEP before sig-node blocked it unless it was dropped from the design.  That being said, a field doesn't need to be a "feature" to be in spec.  It just needs to describe desired state, and have consistent behavior when set.

Every field has to justify its own cost.  If you can implement a feature with a field or without adding API, you should ALMOST ALWAYS choose the option without.

> I don't buy the "excuse not to checkpoint" argument.  There are many things in the API that could have been checkpointed, but that doesn't mean they should be.  NodeName could have been checkpointed in a PV attached to the scheduler, which the scheduler "observes" periodically, and writes to the pod status.

Huh?  NodeName is an actual feature, which is used by people.  And if my time machine was working, I would happily go back and re-apply our current understanding of API semantics to ancient (in k8s terms) decisions.  That isn't useful.  Even if that were a better design "We did it that way before" is the one of the weakest arguments to do it again. All things being equal, defer to precedent, sure.  BUt all things are not equal (to me, yet).

> Or, the priority admission controller could, instead of initializing spec.Priority integer during admission, periodically re-resolve the priority of the priority class, and write it to status.

Again, that was intentional - changing priority at runtime is something we should consider very very carefully.  The design intentionally snapshots the value.

> The question isn't "could this field have been recorded locally in the controller, and written to status?".  The question is "does this field describe desired state" (or what _should_ be), or is this an observation of the thing represented by the object?

"is it state or control" is not the ONLY criteria, or even the primary criteria for being in the API.  Before that is even asked "do we need it in the API" must be evaluated, and that's where I have not been convinced.

I'm saying that the field doesn't allow the user to control anything that is particularly useful and that the suggest feature wherein this field IS useful is itself not self-evidently useful.

Here I am willing to be swayed, but what I am seeing is not super compelling.  Am I just too far away from it to see the value?

>> I still think that proper status signals are needed BUT that doesn't serve as a
>> replacement for checkpoint. The only corner-case is "what if status gets
>> nuked". That doesnt ACTUALLY happen, but we use it as a litms test. Are we
>> going to let that force us down a path we don't know we want?
> 
> The spec vs status litmus test exists for a reason.  It is so that writing controllers is simple and reliable.  They take desired state from a source of truth, observe the actual state of the thing, take corrective actions, and report the actual state as status.  Your example below illustrates my point nicely.  If status is required to know the desired state (i.e. what _should_ be), the above controller strategy no longer works, as the controller would have to ensure state is written to durable storage _before_ taking corrective action.

I think we're crossing streams here.  I'm not saying status is carrying desired state at all.  I am saying `spec.resources` is the only desired state that matters.  If `spec.resources` != `status.resources` AND there's not further status, and controller MUST assume the worst case.

If you're trying to schedule on to that node, you have to assume the larger of those 2 is correct.  If spec > status, that implies that kubelet is about to start the resize".  If spec < status, that implies that kubelet has not yet actuated the downwards-resize.

Unless I am *still* missing something, the only place where this is wrong is when kubelet is not planning to actuate an upwards-resize (i.e. infeasible or similar). That's useful information to surface on its own.

I think the question at hand is, at its core, whether `spec.requiredResources` (fka `spec.resourcesAllocated`) is the best representation of that.  If kubelet changes it before I look at it (NB: it's a race) then I can know "this resize has been accepted".  If kubelet has not updated it, I can only know "kubelet has not updated it".  I'm saying that this is not sufficient, and that patching that deficiency ALSO obviated `spec.requiredResources`.

You are (I think, please correct me) arguing that `spec.requiredResources` is useful on its own.  That's what I am not believing.

> "Down a path we don't know we want" is fair.  I do think my position (field in spec + scheduler involvement) is a reasonable path that we _should_ want.  However, if the kubelet is forever the only controller for resizes, i'm less certain.  I can see arguments for efficiency (no extra write, just included in status), and simplicity of implementation (no subresources, although we now need checkpoints).

I don't buy that this should be part of the scheduler.  To my eyes it is very marginal value for significant complexity.  We're better off having a distinct re-scheduler or something that recognizes bad situations and plays whack-a-mole with them.

> >  t7: kubelet wakes up
> >  t8: kubelet looks at p1, finds status.resources != spec.resources
> >     also finds status.resize = Pending
> > t9; kubelet adds p1 at req=10
> > t10: kubelet adds p2 at req=5
> 
> We probably don't want to do admission based on the contents of pod status for many reasons.  For one, pod status often lags considerably behind what it has admitted.  For example, assume node has 1 core:
> 
> 1. Admit pod A w/ 1000m cpu
> 2. Update pod A status: status.resources.cpu == 1000m
> 2. Resize pod A to 500m cpu.  Accepted by the kubelet!
> 3. Admit pod B w/ 500m cpu

I think the error is here.  You should not schedule it until the resize has been actuated, and you should not admit it.

> 4. Kubelet restarts
> 5. Pod A admitted w/ status.resources.cpu == 1000m
> 6. Pod B rejected w/ request == 500m cpu
> 
> What we can do, which is what we have discussed previously, is checkpoint the "accepted" or "admitted" decisions the kubelet makes.  That way, admission doesn't rely on pod status, just the local file.


That's right, to me.  I don't mean to trivialize the problem of checkpointing, but it's "Not the API's Problem" that kubelet made a decision and then forgot about it.

To recap (trying to make a fair statement here, please point out if I am not even-handed):

1) I think we want clear signal that a change to `spec.resources` has been accepted (ACK or NACK, but not "no-signal").
2) I think we want as little API surface as we can get away with.

* I do not think `spec.requiredResources` is the best representation of (1), and we will need API surface specifically for that.
* You think `spec.requiredResources` enables additional capabilities that justify adding it.
* I am "unconvinced" that the value of `spec.requiredResources` is sufficient (i.e. if fails (2) for me

On to Vinay's response:

@vinaykul:

> - Regarding max(resources, resourcesAllocated) or max(spec, status): It is well intentioned but gives a user some control over how cluster resource is applied. Suppose a node with 4G mem has a guaranteed Pod1 (1G,1G), and evil user starts up a burstable pod P2(1G,-). Now, that user can request 3.1G which will lead to ResizeInfeasable/Deferred condition. Observing that, the user switches back to 1G which would clear that condition, quickly follow it by requesting 3.1G ... and keep toggling to take advantage of durations during which scheduler is not selecting that node. This way it can potentially block off additional 2G with some probability without "paying" for it.

This is a very interesting point.  Your first problem is that you have an evil user... :)

Seriously though, it makes me think.  If this is a criteria, and it seems reasonable, we need to balance it against the risk of scheduler assigning pods to nodes that have other resizes pending that would make the new scheduling decision infeasible.

In other words, it's a bigger problem.  How do you allow a user to request a resource increase without DoS'ing and without "false-feasibility" when everything is fundamentally async.

One path could be to use the max(spec, status) unless no node is feasible, in which case you fall back to the nodes where spec < status (assuming resources will be released) and finally nodes where status < spec.  If we have a proper signal for "infeasible" those nodes could be part of the first set.

> - ```PreferredResources``` (assuming that it is Requests + Limits) sounds like a clean option, let me chew on this a little. Can we use /resize to update spec.Resources if we go with this model? This would avoid needing to do access control in admission plugins, but wouldn't we still need to do access control somewhere outside RBAC on what fields someone is allowed to write - something similar to NodeRestriction plugin to allow node to write only to Resources field.

*IF* we go with 2 spec fields, I think one sub-resource could carry both and we just assert that any controller that can set one can set the other.  That seems DRAMATICALLY better than arbitrary spec access and better than admission control.  We'd want to discuss that with security ppl.  I don't think 2 sub-resources would be justified, especially because the generality of the second one goes way down, but i could be wrong.

So how do we break this impasse, folks?

Do we drag in a 3rd-part API reviewer to weigh in?  Do we drag in a sig-node lead to explain why I am wrong and this "feature" is worthwile?  Both?

I'd like to wrap this topic up - we have more to discuss.

Reply:
I'm resisting the urge to respond to everything... I'll just say that I think the crux of our disagreement is scheduler involvement, not checkpointing (which I have repeatedly said i'm fine with) or users clamoring for the functionality the field provides (which I have said won't be the case).

The path forward depends on what you would accept, so maybe some hypotheticals would help.  What would your response be if...:
1. We went to sig-scheduling and sig-node, and got agreement that having scheduler involvement was an important and useful feature, and that we wanted the scheduler making resource allocation decisions.
2. We fell back to our plan B, which is checkpointing resourcesAllocated, and moving the field as-is to pod status.

Or we can try to come up with a new design that probably relies on a checkpoint, and exposes different information in status.

Reply:
At sig-node today, we reached consensus on the second option.  So if that works for you, we can move forward.  Derek did point out an issue that we will have to figure out an answer for, but both Dawn and Derek were OK with the switch.

Reply:
@thockin FYI, here's [overview](https://drive.google.com/file/d/12wKXl6_y0tDqOhTEXPIK7gN_qsfJzjJ9/view) of the node checkpointing based approach (@dashpole 's second option above that we discussed in sig-node this morning). 

I'll be working on the details of leveraging the node checkpointing framework and defining what checkpoint state (internal to kubelet) looks like.

Reply:
I am sorry I missed this update!!  I was waiting for an email...sigh.

I looked at the overview, and I think it looks good.  We should consider the problem you described (flapping to deny the scheduler access) and decide to either do it or file an issue for later evaluation.  The workaround I proposed is crude, but maybe enough.

Let's focus on the other areas now :)

Reply:
> I am sorry I missed this update!! I was waiting for an email...sigh.
> 
> I looked at the overview, and I think it looks good. We should consider the problem you described (flapping to deny the scheduler access) and decide to either do it or file an issue for later evaluation. The workaround I proposed is crude, but maybe enough.
> 
> Let's focus on the other areas now :)

Great, thanks for reviewing the doc! 

I should be able to give some time to this starting Sept.  I'll do a new PR taking in much of the stuff already implemented in [PR 92127](https://github.com/kubernetes/kubernetes/pull/92127) and apply this API modification that moves ResourcesAllocated from Spec to Status. I don't expect it to be a major change requiring me to retest a lot of stuff, except for the new checkpointing code in kubelet.

For the flapping issue, the risk is low since VPA is primary consumer, and we don't expect it to indulge in any funny business. So I'm leaning towards filing a tracking issue and note it down as a risk in the KEP.